python/llm/src/bigdl/llm/transformers/models/llama.py

```diff
@@ -334,108 +334,141 @@ def llama_attention_selective_batching_forward_4_31(
-    #                                                                      self.q_proj.weight.qtype,
-    #                                                                      kv_seq_len,
-    #                                                                      self.head_dim)
-    #     kv_seq_len += 1
-
-    # else:
-    if self.config.pretraining_tp > 1:
-        invalidInputError(False, f"vLLM: config.pretraining_tp > 1 not supported yet")
+    if decoding_fast_path:
+        past_k = past_key_value[0][0]
+        past_v = past_key_value[0][1]
+        kv_seq_len = past_k.shape[-2]
+        if not enough_kv_room:
+            new_cache_k, new_cache_v = extend_kv_cache(1,
+                                                       self.num_key_value_heads,  # Support GQA
+                                                       self.head_dim,
+                                                       kv_seq_len,
+                                                       kv_seq_len +
+                                                       KV_CACHE_ALLOC_BLOCK_LENGTH,
+                                                       dtype=past_k.dtype,
+                                                       device=device)
+            new_cache_k[:] = past_k
+            new_cache_v[:] = past_v
+            past_k = new_cache_k
+            past_v = new_cache_v
+        hidden_states = hidden_states.view(1, -1)
+        import linear_q4_0
+        query_states, key_states, value_states = linear_q4_0.forward_qkv(hidden_states,
+                                                                         self.q_proj.weight,
+                                                                         self.k_proj.weight,
+                                                                         self.v_proj.weight,
+                                                                         position_ids,
+                                                                         past_k, past_v,
+                                                                         self.q_proj.weight.qtype,
+                                                                         kv_seq_len,
+                                                                         self.head_dim)
+        kv_seq_len += 1
     else:
-        query_states = self.q_proj(hidden_states)
-        key_states = self.k_proj(hidden_states)
-        value_states = self.v_proj(hidden_states)
-
-    query_states = query_states.view(bsz, q_len,
-                                     self.num_heads, self.head_dim).transpose(1, 2)
-    key_states = key_states.view(bsz, q_len,
-                                 self.num_key_value_heads, self.head_dim).transpose(1, 2)
-    value_states = value_states.view(bsz, q_len,
+        if self.config.pretraining_tp > 1:
+            invalidInputError(False, f"vLLM: config.pretraining_tp > 1 not supported yet")
```

Comment:
why the error message is `vLLM` specific?

Reply:
We have not yet consider the case where `self.config.pretraining_tp` greater than one.  I think the original `llama_attention_forward_4_31` handles this case correctly.  So this error message should be `vLLM` specific.